### ISSUE TITLE
Move default views to their respective modules

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#361]
As per https://github.com/backdrop/backdrop-issues/issues/147#issuecomment-58574317, default views should be provided by their module and not Views itself.

---
~~PR by @BWPanda: https://github.com/backdrop/backdrop/pull/2391~~
PR by @jenlampton https://github.com/backdrop/backdrop/pull/2464

**Comment by BWPanda** [comment]
It's been a long time coming, but I just made a new PR for this issue (no, I haven't been working on it all these years :stuck_out_tongue_closed_eyes:).

**Comment by klonos** [comment]
...I have reviewed the PR and it seem good code-wise, but when testing in the PR sandbox I have noticed that the "Recent comments" view was still available even when the Comments module was disabled **and uninstalled**.

I would expect that if a module that provides a view is disabled, then it should not be possible to enable the view.

I would also expect that if the module is **also uninstalled**, then the view provided by that module should not be listed in the Views UI at all.

**Comment by klonos** [comment]
...I forgot to mention that before disabling and uninstalling the module, I had also reverted the view.

**Comment by BWPanda** [comment]
> when testing in the PR sandbox I have noticed that the "Recent comments" view was still available even when the Comments module was disabled and uninstalled.

I noticed the same. I think it may have to do with the 'storage' config option not always being available...?

**Comment by BWPanda** [comment]
Based on this comment in `view.inc`:

> Storage settings are only saved if this view was provided by a module.
> Otherwise VIEWS_STORAGE_NORMAL is assumed on load.

I've gone ahead and changed the PR to delete views on uninstall if the view has not been overridden (as opposed to if it's in default state). This is because, as far as I can tell, unless a module-provided view has been overridden, the 'storage' option may not exist. We can therefore assume that it's in its default state, since we know it's a module-provided view (as opposed to a user-created one, which would have a 'normal' state).

In other words, I simply changed `if ($config->get('storage') == VIEWS_STORAGE_DEFAULT) {` to `if ($config->get('storage') != VIEWS_STORAGE_OVERRIDE) {` and the view is now properly deleted on module uninstall.

**Comment by klonos** [comment]
Not sure that this works as expected quite yet. I have tested this in the PR sandbox:

1. Visited the modules listing page and confirmed that the Comment module is enabled.
2. Visited the views listing page and confirmed that the "Recent comments" view was included. It was disabled.
3. I reverted the "Recent comments" view (to make sure that it is in its "default" state).
4. I **disabled** the Comment module.
5. The "Recent comments" view was still listed.
6. Tried to enabled the "Recent comments" view -> it got enabled 👎 ...I expect views supplied by modules to not be possible to be enabled if the respective module is disabled - either disable their "Enable" button, or require the module to be enabled in order to enable the view (separate ticket perhaps)
7. I **uninstalled** the Comment module.
8. Visited the view listing page -> the "Recent comments" view was still listed, and in fact enabled 👎 ...it's OK for views supplied by modules to be listed even if the module is disabled or uninstalled (so long as the module code is present in the codebase - which is always true for core modules), but to have the view enabled while the respective module is disabled and uninstalled seems really odd.

**Comment by BWPanda** [comment]
I agree that we can do things better regarding disabled modules, etc., but I wonder what the scope of this issue is... If it's simply to move the views out of Views and into their respective modules, then maybe we should create a separate issue for @klonos' suggestions. If we want this issue to be inclusive of those suggestions though, I'll try looking at this later to see what can be done.

**Comment by klonos** [comment]
I am a fan of "baby steps" ...so we can have a separate issue for improving the handling of views provided by modules. That's why in my initial review comment a few days back I said:

> I have reviewed the PR and it seems good code-wise ...

It was my way of saying that the PR is RTBC for its scope (moving the code). The other things I have reported are pre-existing issues anyway.

**Comment by quicksketch** [comment]
We should fix the problem @klonos described above, where Comment and Taxonomy modules will only take ownership of their views on new installs instead of existing ones. We need an update hook in both modules that just loads in their corresponding view config files, updates the `module` key, and then saves it.

**Comment by klonos** [comment]
Setting to NW according to @quicksketch's comment above.

**Comment by jenlampton** [comment]
> We need an update hook in both modules that just loads in their corresponding view config files, updates the module key, and then saves it.

I added these yesterday, the latest PR has them. Needs review: https://github.com/backdrop/backdrop/pull/2464

**Comment by herbdool** [comment]
Looks good to me.

**Comment by herbdool** [comment]
One question, should we include views.view.promoted in this PR?

**Comment by jenlampton** [comment]
Into Node module?

**Comment by herbdool** [comment]
I think yes

**Comment by quicksketch** [comment]
It does seem weird that `views.view.promoted` stays within the Views module instead of Node. Could we apply this same change to that view as well?

**Comment by jenlampton** [comment]
I've moved the promoted view to the node module, and added an update hook to switch the module from `views` to `node`.

**Comment by quicksketch** [comment]
Looks like tests aren't passing because of views_install() referencing the promoted view config file.

**Comment by jenlampton** [comment]
Do you think I need to move the old update hook?

**Comment by jenlampton** [comment]
Hm, Views install doesn't reference the old config file: 
```
/**
 * Implements hook_install().
 */
function views_install() {
  db_query("UPDATE {system} SET weight = 10 WHERE name = 'views'");
}
```

**Comment by quicksketch** [comment]
> I'll try changing the module to node, but is it okay to leave this update in views module?

Yes. Though we shouldn't actually be doing this at all. This is why we copy the entire view into update hooks, and why we have an issue at #3347. Update hooks shouldn't be dependent upon the default configuration of a module, because that default configuration changes over time.

**Comment by quicksketch** [comment]
Thanks @jenlampton for the quick update! I've merged https://github.com/backdrop/backdrop/pull/2464 into 1.x for 1.13.0. Thanks @BWPanda, @klonos, and @herbdool for your work here!
